src/librustc_trans/type_.rs

```diff
@@ -297,4 +298,38 @@ impl Type {
     pub fn x86_mmx(cx: &CodegenCx) -> Type {
         ty!(llvm::LLVMX86MMXTypeInContext(cx.llcx))
     }
+
+    pub fn va_list(cx: &CodegenCx, name: &str) -> Type {
+        let int_t = Type::c_int(cx);
+        let voidp_t = Type::i8(cx).ptr_to();
+        match cx.tcx.sess.target.target.options.va_list_kind {
+            VaListKind::CharPtr => {
+                Type::i8p(cx)
+            },
+            VaListKind::VoidPtr => {
+                voidp_t
+            }
+            VaListKind::X86_64Abi => {
+                let mut va_list_t = Type::named_struct(cx, name);
+                va_list_t.set_struct_body(&[int_t, int_t, voidp_t, voidp_t], false);
+                va_list_t
+            },
+            VaListKind::AArch64Abi => {
+                let mut va_list_t = Type::named_struct(cx, name);
+                va_list_t.set_struct_body(&[voidp_t, voidp_t, voidp_t, int_t, int_t], false);
+                va_list_t
+            },
+            VaListKind::PowerPcAbi => {
+                let i8_t = Type::i8(cx);
+                let i16_t = Type::i16(cx);
+                let mut va_list_t = Type::named_struct(cx, name);
+                va_list_t.set_struct_body(&[i8_t, i8_t, i16_t, voidp_t, voidp_t], false);
+                va_list_t
+            },
+            _ => {
+                // TODO: Is this the right thing to do?
+                bug!("va_list: not supported for this target type");
```

Comment:
Yes, `bug`-ing out here is the right thing to do as the valid variants are filtered out earlier. Message could be made to point out the issue better, though:

> unexpected va_list kind reached trans

for example.

Reply:
:+1: much better. I'll do a bit more research, but I think defaulting to either a void pointer or char pointer would be reasonable.